### PR TITLE
Better checking if two results are equal

### PR DESCRIPTION
**Pull request by darcythomas**
When creating an experiment a Func<T,T,bool>  or  an instance of IEqualityComparer<T> can be passed in.

I have also refactored the equality checking code into its own class and written unit tests (including checks for if exceptions are equal).

Spec:

``` cs
        /// <summary>
        /// Checks if two ExperimentResults are equal
        /// </summary>
        /// <param name="controlResult">Control ExperimentResult</param>
        /// <param name="candidateResult">Candidate ExperimentResult</param>
        /// <returns>
        ///  Returns true if: 
        /// 
        /// 
        ///  The values of the observations are equal according to an IEqualityComparer&lt;T&gt; expression, if given 
        ///  The values of the observations are equal according to a comparison function, if given  
        ///  Both observations raised an exception with the same Type and message.
        ///  The values of the observation are both null 
        ///  The values of the observations are equal according to Ts IEquatable&lt;T&gt; implementation, if implemented 
        ///  The values of the observations are equal (using .Equals()) 
        ///  
        ///  Returns false otherwise. 
        /// </returns>
```

### CODE REVIEW

File: src/Scientist/Internals/Experiment.cs [hunk 1]
```diff
@@ -48,14 +74,8 @@ public ExperimentInstance(string name, Func<Task<T>> control, Func<Task<T>> cand
                 controlResult = await Run(_control);
             }
 
-            // TODO: We need to compare that thrown exceptions are equivalent too https://github.com/github/scientist/blob/master/lib/scientist/observation.rb#L76
-            // TODO: We're going to have to be a bit more sophisticated about this.
-            bool success =
-                controlResult.Result == null && candidateResult.Result == null
-                || controlResult.Result != null && controlResult.Result.Equals(candidateResult.Result)
-                || controlResult.Result == null && candidateResult.Result != null;
+            bool success = _experimentResultComparer.Equals(controlResult, candidateResult);
```
Comment (darcythomas) [inline src/Scientist/Internals/Experiment.cs]: Use of new ExperimentResultComparer

File: src/Scientist/Internals/Experiment.cs [hunk 2]
```diff
@@ -48,14 +74,8 @@ public ExperimentInstance(string name, Func<Task<T>> control, Func<Task<T>> cand
                 controlResult = await Run(_control);
             }
 
-            // TODO: We need to compare that thrown exceptions are equivalent too https://github.com/github/scientist/blob/master/lib/scientist/observation.rb#L76
```
Comment (darcythomas) [inline src/Scientist/Internals/Experiment.cs]: Refactored this out and implemented for checking of exceptions as well.

File: UnitTests/ComplexResult.cs [hunk 3]
```diff
@@ -0,0 +1,15 @@
+﻿using System;
+
+public class ComplexResult : IEquatable<ComplexResult>
```
Comment (darcythomas) [inline UnitTests/ComplexResult.cs]: Class implementing IEquatable

File: UnitTests/ExperimentResultComparerTests.cs [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+using System.Text;
+using System.Threading.Tasks;
+using GitHub;
+using GitHub.Internals;
+using NSubstitute;
+using Xunit;
+
+namespace UnitTests
+{
+    public class ExperimentResultComparerTests
```
Comment (darcythomas) [inline UnitTests/ExperimentResultComparerTests.cs]: many tests

File: UnitTests/ScientistTests.cs [hunk 5]
```diff
@@ -121,5 +122,103 @@ public void AnExceptionReportsDuration()
             Assert.True(((InMemoryObservationPublisher)Scientist.ObservationPublisher).Observations.First(m => m.Name == "failure").ControlDuration.Ticks > 0);
             Assert.True(((InMemoryObservationPublisher)Scientist.ObservationPublisher).Observations.First(m => m.Name == "failure").CandidateDuration.Ticks > 0);
         }
+
```
Comment (darcythomas) [inline UnitTests/ScientistTests.cs]: This came in from a merge...

File: src/Scientist/Internals/ExperimentResultComparer.cs [hunk 6]
```diff
@@ -0,0 +1,108 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace GitHub.Internals
+{
+    internal class ExperimentResultComparer<T> : IEqualityComparer<ExperimentInstance<T>.ExperimentResult>
```
Comment (darcythomas) [inline src/Scientist/Internals/ExperimentResultComparer.cs]: Shiny new class to do all the result equality checking.

File: src/Scientist/Internals/ExperimentResultComparer.cs [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+        private readonly IEqualityComparer<T> _resultEqualityComparer;
+
+        public ExperimentResultComparer(IEqualityComparer<T> resultEqualityComparer, Func<T, T, bool> resultComparison)
+        {
+            _resultEqualityComparer = resultEqualityComparer;
+            _resultComparison = resultComparison;
+        }
+
+
+        /// <summary>
```
Comment (darcythomas) [inline src/Scientist/Internals/ExperimentResultComparer.cs]: Woot a spec

File: src/Scientist/Internals/ExperimentResultComparer.cs [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+        ///  Both observations raised an exception with the same Type and message.
+        ///  The values of the observation are both null 
+        ///  The values of the observations are equal according to Ts IEquatable&lt;T&gt; implementation, if implemented 
+        ///  The values of the observations are equal (using .Equals()) 
+        ///  
+        ///  Returns false otherwise. 
+        /// </returns>
+        public bool Equals(ExperimentInstance<T>.ExperimentResult controlResult, ExperimentInstance<T>.ExperimentResult candidateResult)
+        {
+            if (_resultComparison != null)
```
Comment (darcythomas) [inline src/Scientist/Internals/ExperimentResultComparer.cs]: I would like to refactor this a bit more but this is good enough for now.
I don't like having a whole bunch of `if(){return xyz()}`
(I would refactor into some sort of strategy pattern where all tests have to not fail or something)

**Comment by Haacked** [comment]
I think a previous PR made this change unnecessary.

**Pull request closed without merging**